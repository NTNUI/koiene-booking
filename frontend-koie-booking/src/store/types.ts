// rootState = generic type, defines the root state type
import AdminBooking, { AdminBookingDictionary } from '@/types/admin/AdminBooking';
import { AdminReportKoieDictionary } from '@/types/admin/AdminReports';

export interface RootState {
  version: string;
  koie: KoieState;
  booking: BookingState;
  auth: AuthState;
  avalanche: AvalancheState;
  adminBookings: AdminBookingsState;
}

export interface KoieState {
  currentLocale: string;
  isLoading: boolean;
  koieData: Koie;
  error: boolean;
  allKoier: [];
}

export interface Koie {
  location: {
    latitude: string;
    longitude: string;
    area: string;
    difficultyInfo_eng: string;
    difficultyInfo_nor: string;
    altitude: number;
    terrain_eng: string;
    terrain_nor: string;
    map_pdf: string;
    kartblad: string;
  };
  description: {
    yr_link_eng: string;
    yr_link_nor: string;
    description_eng: string;
    description_nor: string;
    directions_eng: string;
    directions_nor: string;
    parking_eng: string;
    parking_nor: string;
  };
  beds_available_in_booking_window: Array<BedsAvailable>;
  name: string;
  number_of_beds: number;
  booking_window: number;
  price_member: number;
  price_not_member: number;
}

export interface BedsAvailable {
  date: string;
  bedsAvailable: number;
}

export interface BookingState {
  isLoading: boolean;
  dateFrom: string;
  dateTo: string;
  beds: number;
  step: number;
  guests: Array<Guest>;
  numberOfMembers: number;
  numberOfNonMembers: number;
  edited: boolean;
  validForm: boolean;
  unavailableDates: [];
  availableBeds: number;
  noDatesAvailable: boolean;
  error: boolean;
  bookingData: BookingInfo;
  paymentConfirmed: boolean;
}

export interface ReportState {
  validForm: boolean;
  edited: boolean;
  isLoading: boolean;
  error: boolean;
  step: number;
  reportData: ReportData;
  bookingInfo: BookingInfo;
}

export interface ReportData {
  id: number;
  booking_uuid: string;
  firewood: number;
  chopped_up_wood: number;
  smoke_detector_is_working: boolean;
  gas_is_full: boolean;
  gas_burner_primus: number;
  axe: number;
  hammer: number;
  saw: number;
  saw_blade: number;
  saw_bench: number;
  spade: number;
  kerosene_lamp: number;
  detergent: number;
  dishware: number;
  cookware: number;
  cabin_book: number;
  candle_holders: number;
  fire_blanket: number;
  fire_extinguisher: number;
  other_faults: string;
  boat_status: number;
  canoe_status: number;
  life_jackets_status: number;
  feedback: string;
}

export interface AvalancheState {
  warningLevels: Array<Warning>;
  disableBooking: boolean;
  warningData: Array<WarningResult>;
  koierToCheck: Array<string>;
}

export interface Warning {
  day: string;
  level: number;
}

export interface WarningResult {
  dangerLevel: number;
  validTo: string;
}

export interface CreateBookingInfo {
  koie: string;
  from_date: string;
  to_date: string;
  guests: Array<Guest>;
  guests_member: number;
  guests_not_member: number;
}

export interface BookingInfo {
  id: number;
  uuid: string;
  user: number;
  koie: string;
  price: number;
  booking_transaction_id: string;
  from_date: string;
  to_date: string;
  guests_member: number;
  guests_not_member: number;
}

export interface AuthState {
  isAdmin: boolean;
  tokens: {
    access: string | null;
    refresh: string | null;
  };
}

export interface Guest {
  name: string;
  number: string;
  email: string;
  isMember: boolean;
}

export interface AdminBookingsState {
  cabinsWithBookings: AdminBookingDictionary;
  startDate: string;
}

export interface AdminReportsState {
  chosenCabin: string;
  expanded: boolean;
  cabins: AdminReportKoieDictionary;
}
