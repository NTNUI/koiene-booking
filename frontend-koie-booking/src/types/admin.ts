import { Component } from 'vue';

export interface AdminView {
  id: String; // Used to reference the buttons in SideBar
  title: String;
  icon: String;
  component: Component;
}
