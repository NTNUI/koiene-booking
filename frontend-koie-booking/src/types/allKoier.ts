export interface AllKoierData {
  koieTitle: string;
}
