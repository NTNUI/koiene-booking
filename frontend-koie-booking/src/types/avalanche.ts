import { TranslateResult } from 'vue-i18n';

export interface AvalancheData {
  result: string;
  levelFirstDay: string;
  levelSecondDay: string;
  levelThirdDay: string;
  firstDay: string;
  secondDay: string;
  thirdDay: string;
  fromDate: string;
  nextDate: string;
  lastDate: string;
  warningTypes: Array<TranslateResult>;
  colors: Array<string>;
}
