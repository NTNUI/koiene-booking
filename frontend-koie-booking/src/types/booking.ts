import { TranslateResult } from 'vue-i18n';

export interface BookingData {
  step: number;
  labels: Array<TranslateResult>;
}

export interface BookingFirstStepData {
  bedsTotal: number;
  bedsChosen: number;
  error: boolean;
}

export interface BookingSecondStepData {
  bedsChosen: number;
  bookingStep: number;
  guests: Array<Guest>;
  numberRules: object;
  nameRules: object;
  emailStrictRules: object;
  emailLooseRules: object;
  edited: boolean;
  validForm: boolean;
  memberPrice: number;
  nonMemberPrice: number;
}

export interface BookingThirdStepData {
  complete: boolean;
  stripeOptions: object;
  totalPrice: number;
}

export interface Guest {
  name: string;
  number: string;
  isMember: boolean;
  email?: string;
}
