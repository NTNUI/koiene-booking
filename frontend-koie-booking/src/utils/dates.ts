import dayjs from 'dayjs';
require('dayjs/locale/nb');
dayjs.locale('nb');

export function getDateString(baseDate?: string, addDays?: number): string {
  let resultString: string;
  if (baseDate && addDays) {
    resultString = new Date(new Date(baseDate).getTime() + addDays * 24 * 60 * 60 * 1000).toISOString().substr(0, 10);
  } else if (baseDate) {
    resultString = new Date(baseDate).toISOString().substr(0, 10);
  } else if (addDays) {
    resultString = new Date(new Date().getTime() + addDays * 24 * 60 * 60 * 1000).toISOString().substr(0, 10);
  } else {
    resultString = new Date().toISOString().substr(0, 10);
  }
  return resultString;
}

export function getDate(baseDate?: string, addDays?: number): Date {
  let resultDate: Date;
  if (baseDate && addDays) {
    resultDate = new Date(new Date(baseDate).getTime() + addDays * 24 * 60 * 60 * 1000);
  } else if (baseDate) {
    resultDate = new Date(baseDate);
  } else if (addDays) {
    resultDate = new Date(new Date().getTime() + addDays * 24 * 60 * 60 * 1000);
  } else {
    resultDate = new Date();
  }
  return resultDate;
}

export function addToDate(startDate: string, howMany: number, timeUnit: dayjs.OpUnitType): string {
  const date = dayjs(startDate + 'T00:00:00.000Z');
  return date.add(howMany, timeUnit).format('YYYY-MM-DD');
}

export function formatDate(dateISO: string, formatString: string): string {
  const date = dayjs(dateISO + 'T00:00:00.000Z');
  return date.format(formatString);
}
