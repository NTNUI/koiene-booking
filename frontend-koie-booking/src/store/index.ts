import Vue from 'vue';
import Vuex, { StoreOptions } from 'vuex';
import axios from 'axios';
import VueAxios from 'vue-axios';
import { RootState } from './types';
import { koie } from './modules/users/koie/index';
import { booking } from './modules/users/booking/index';
import { avalanche } from './modules/users/avalanche/index';
import { report } from './modules/users/report/index';
import { auth } from './modules/users/auth/index';
import { adminBookings } from '@/store/modules/admin/bookings';

Vue.use(Vuex, VueAxios, axios);

const store: StoreOptions<RootState> = {
  modules: {
    adminBookings,
    koie,
    booking,
    avalanche,
    report,
    auth
  }
};

export default new Vuex.Store<RootState>(store);

export { store as storeConfig };
