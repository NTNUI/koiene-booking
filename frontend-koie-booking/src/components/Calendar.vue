<template>
  <v-container :class="row ? $style.row : $style.column" v-bind="$props">
    <v-menu
      ref="menu1"
      v-model="menu1"
      :close-on-content-click="false"
      transition="scale-transition"
      offset-y
      max-width="290px"
      min-width="290px"
      dark
    >
      <template v-slot:activator="{ on }" :class="$style.templateTextField">
        <v-text-field
          v-model="computedDateFormattedFrom"
          :label="$t('calendar.arrival')"
          prepend-icon="event"
          dark
          :disabled="noDatesAvailable"
          :class="row && $style.rowTextFieldArrival"
          @blur="dateFrom = parseDate(dateFormatted(temporaryDateFrom))"
          v-on="on"
        ></v-text-field>
      </template>
      <v-date-picker
        v-model="dateFrom"
        :min="minimumDate"
        :max="maximumDate"
        :color="this.$scssVars.globalColorPrimary"
        :allowed-dates="allowedDatesFrom"
        no-title
        @input="menu1 = false"
        @change="updateDisabledDatesInMenu"
      ></v-date-picker>
    </v-menu>
    <v-menu
      ref="menu2"
      v-model="menu2"
      :close-on-content-click="false"
      transition="scale-transition"
      offset-y
      max-width="290px"
      min-width="290px"
      dark
    >
      <template v-slot:activator="{ on }" :class="$style.templateTextField">
        <v-text-field
          v-model="computedDateFormattedTo"
          :label="$t('calendar.departure')"
          prepend-icon="event"
          :disabled="noDatesAvailable"
          dark
          :class="[row ? $style.rowTextFieldDeparture : '', row && !allkoier ? $style.bookingTabletRemovePadding : '']"
          @blur="
            {
              dateTo = parseDate(dateFormatted(temporaryDateTo));
              updatedDateFrom = !updatedDateFrom;
            }
          "
          v-on="on"
        ></v-text-field>
      </template>
      <v-date-picker
        v-model="dateTo"
        :min="minimumDateTo"
        :max="maximumDateTo"
        no-title
        :color="this.$scssVars.globalColorPrimary"
        @input="menu2 = false"
        @change="updateDisabledDatesInMenu"
      ></v-date-picker>
    </v-menu>
    <div :class="$style.divBeds">
      <p v-if="vbeds" :class="$style.bedsText" v-bind="$props">
        {{ $t('calendar.beds') }}
      </p>
      <p v-if="!row" :class="$style.availableBedsText">( {{ availableBeds }} available )</p>
      <v-text-field
        v-if="vbeds"
        v-model="beds"
        :value="beds"
        :class="$style.bedsInput"
        v-bind="$props"
        min="1"
        :max="availableBeds"
        :disabled="noDatesAvailable"
        :dense="true"
        dark
        hide-details
        single-line
        outlined
        type="number"
        @change="(e) => setBeds(e)"
      />
    </div>
    <v-alert
      v-if="error"
      v-model="error"
      transition="expand-transition"
      :dismissible="true"
      :class="$style.errorMsg"
      type="info"
      @click="error = false"
    >
      {{ $t('calendar.errormsg') }} {{ availableBeds }}
    </v-alert>
  </v-container>
</template>

<script lang="ts">
import { CalendarData } from '../types/calendar';
import Vue from 'vue';
import { getDateString, getDate } from '@/utils/dates';

export default Vue.extend({
  name: 'Calendar',
  props: {
    row: { type: Boolean, default: true },
    vbeds: { type: Boolean, default: true },
    allkoier: { type: Boolean, default: false }
  },
  data(): CalendarData {
    return {
      menu1: false,
      menu2: false,
      beds: this.$store.state.booking.beds,
      disabledDates: [],
      minimumDate: getDateString(),
      maximumDate: getDateString(),
      maximumDateTo: getDateString(),
      error: false,
      updatedDateFrom: false,
      temporaryDateTo: this.$store.state.booking.dateTo || getDateString(undefined, 1),
      temporaryDateFrom: this.$store.state.booking.dateFrom || getDateString(),
      setDatesFinished: false
    };
  },
  computed: {
    computedDateFormattedFrom(): string | null {
      return this.formatDate(this.dateFrom);
    },
    computedDateFormattedTo(): string | null {
      return this.formatDate(this.dateTo);
    },
    minimumDateTo(): string {
      return getDateString(this.dateFrom, 1);
    },
    daysInAdvanceBooking(): number {
      return this.$store.state.koie.koieData.booking_window || 14;
    },
    availableBeds(): number {
      if (this.allkoier) {
        this.$store.dispatch('booking/SET_AVAILABLE_BEDS', 25);
      }
      if (this.$store.state.booking.availableBeds != undefined) {
        return this.$store.state.booking.availableBeds;
      }
      return 10;
    },
    totalBeds(): number {
      return this.$store.state.koie.koieData.number_of_beds;
    },
    warningLevels(): Array<any> {
      return this.$store.state.avalanche.warningLevels;
    },
    noDatesAvailable(): boolean {
      return this.$store.state.booking.noDatesAvailable;
    },
    koieTitle(): string {
      return this.$store.state.koie.koieData.name;
    },
    dateFrom: {
      get(): string {
        return this.temporaryDateFrom;
      },
      set(newValue: string) {
        if (newValue === '--------') {
          newValue = this.dateFrom;
        }
        this.temporaryDateFrom = newValue;
        if (getDate(this.temporaryDateTo) <= getDate(this.dateFrom)) {
          this.dateTo = getDateString(newValue, 1);
        }
        if (this.setDatesFinished) {
          this.updateDisabledDatesInMenu();
          this.$store.dispatch('booking/SET_DATE_FROM', this.dateFrom);
          this.getAvailableBeds();
          this.setBeds(this.beds.toString());
        }
      }
    },

    dateTo: {
      get(): string {
        return this.temporaryDateTo;
      },
      set(newValue: string) {
        this.temporaryDateTo = newValue;
        if (this.setDatesFinished) {
          this.updateDisabledDatesInMenu();
          this.$store.dispatch('booking/SET_DATE_TO', this.temporaryDateTo);
          this.getAvailableBeds();
          //setting beds here makes sure that if you choose a date where only 9 beds is available and you have already chosen 10 beds, then beds is reset to 1
          this.setBeds(this.beds.toString());
        }
      }
    }
  },
  mounted() {
    this.maximumDate = getDateString(undefined, this.daysInAdvanceBooking - 1);
    this.maximumDateTo = getDateString(undefined, this.daysInAdvanceBooking);
    this.minimumDate = getDateString();

    this.setDates();
    this.updateDisabledDatesInMenu();
    //checks if we are on AllKoier.vue, there's no maximumBeds since we look at all the cabins
    if (this.$props.allkoier === false) {
      this.getAvailableBeds();
    }
    if (this.$store.state.booking.beds <= this.availableBeds) {
      this.beds = this.$store.state.booking.beds;
    }
    if (this.$store.state.booking.dateFrom.length === 0) {
      //if the dates are not sat, we set the date in store to be todays date so it won't be undefined
      this.$store.dispatch('booking/SET_DATE_FROM', this.dateFrom);
      this.$store.dispatch('booking/SET_DATE_TO', this.dateTo);
    }
  },
  methods: {
    disableBooking() {
      if (this.$store.state.avalanche.koierToCheck.includes(this.koieTitle)) {
        for (let e in this.warningLevels) {
          if (this.warningLevels[e].level > 2) {
            this.$store.dispatch('avalanche/DISABLE_BOOKING', true);
            return;
          } else {
            this.$store.dispatch('avalanche/DISABLE_BOOKING', false);
          }
        }
      }
    },
    formatDate(date: string): string | null {
      if (!date) return null;
      const [year, month, day] = date.split('-');
      let dateString: string = '';
      if (this.noDatesAvailable) {
        dateString = '--/--/--';
      } else {
        dateString = `${month}/${day}/${year}`;
      }
      return dateString;
    },
    parseDate(date: string | null): string | null {
      if (date === '--------' || date === '--/--/--') {
        date = this.formatDate(this.dateFrom);
      }
      if (!date) return null;
      let parsedDate = '';
      if (date.includes('/')) {
        const [month, day, year] = date.split('/');
        parsedDate = `${year}-${month.padStart(2, '0')}-${day.padStart(2, '0')}`;
      } else {
        parsedDate = date;
      }
      return parsedDate;
    },
    setBeds(number: string) {
      if (parseInt(number) <= this.availableBeds) {
        this.$store.dispatch('booking/SET_BEDS', number);
        this.checkNoDatesAvailable();
      } else {
        this.error = true;
        this.beds = 1;
        this.$store.dispatch('booking/SET_BEDS', 1);
      }
    },
    allowedDatesFrom(val: string): string {
      let allowedDate: string = '';
      if (!this.disabledDates.includes(val)) {
        allowedDate = val;
      }
      return allowedDate;
    },
    dateFormatted(date: string): string | null {
      if (date === '--/--/--' || date === '--------') {
        date = this.dateFrom;
      }
      return getDateString(date);
    },
    async getAvailableBeds() {
      const bedsAvailableBetweenDates: number[] = this.bedsAvailableBetweenDates(this.dateFrom, this.dateTo);
      const available = Math.min.apply(null, bedsAvailableBetweenDates);
      this.checkNoDatesAvailable(available);
      if (available) {
        await this.$store.dispatch('booking/SET_AVAILABLE_BEDS', available);
      }
    },
    //find available beds per day, between two dates at a given cabin. Also update disabledDates list.
    bedsAvailableBetweenDates(dateFrom: string, dateTo: string): number[] {
      try {
        const listOfDates = [];
        listOfDates.push(dateFrom);
        let tempDate = dateFrom;
        //add all the dates between fromdate and todate to a list
        const dateToMinus1 = getDate(dateTo, -1);
        while (getDate(tempDate) < dateToMinus1) {
          tempDate = getDateString(tempDate, +1);
          listOfDates.push(tempDate);
        }
        const bedsAvailablePerDate = listOfDates.map((date) => {
          if (
            this.$store.state.koie.koieData.beds_available_in_booking_window[date] < 1 ||
            this.$store.state.koie.koieData.beds_available_in_booking_window[date] < this.beds
          ) {
            this.disabledDates.push(date);
          } else {
            //if we change something so that the date should be possible to choose, we have to remove it from disableddates array
            if (this.disabledDates.includes(date)) {
              const index = this.disabledDates.indexOf(date);
              if (index > -1) {
                this.disabledDates.splice(index, 1);
              }
            }
          }
          return this.$store.state.koie.koieData.beds_available_in_booking_window[date];
        });
        return bedsAvailablePerDate;
      } catch (err) {
        if (process.env.NODE_ENV !== 'test') {
          throw err.message;
        }
        return [];
      }
    },
    setDates() {
      /*checks if todays date is in the disabled dates, in that case, sets dateFrom to be the next date, then
      checks that date as well. If the dateFrom gets set to the date after maximumDate, it means all bookable dates are unavailable.
      */
      // Initialize disabledDatesList
      this.bedsAvailableBetweenDates(this.minimumDate, this.maximumDate);

      this.dateFrom = this.$store.state.booking.dateFrom || getDateString();

      // If the store dateFrom is initially beyond the booking window of this koie, reset the dates.
      const resetDateTo = this.checkDatesBeyondBookingWindow();

      let dateFromUpdated = false;
      if (this.disabledDates.includes(this.dateFrom) || this.maximumDate === this.dateFrom) {
        this.setDatesFinished = false;
        dateFromUpdated = true;
        while (this.disabledDates.includes(this.dateFrom)) {
          this.dateFrom = getDateString(this.dateFrom, 1);
          if (this.dateFrom === this.maximumDate) {
            break;
          }
        }
      }

      this.checkNoDatesAvailable();
      this.setDatesFinished = true;
      // To trigger the dateFrom.set function
      this.dateFrom = this.dateFrom;

      // If dateFrom has been updated because of unavailability or store value is empty, then temporaryDateTo = temporaryDateFrom + 1
      if (dateFromUpdated || resetDateTo || !this.$store.state.booking.dateTo) {
        this.dateTo = getDateString(this.dateFrom, 1);
      } else {
        this.dateTo = this.$store.state.booking.dateTo;
      }
    },
    checkNoDatesAvailable(available = 5) {
      if (!this.allkoier && (getDate(this.dateFrom) > getDate(this.maximumDate) || available < 1)) {
        this.$store.dispatch('booking/SET_NO_DATES_AVAILABLE', true);
        this.$store.dispatch('avalanche/DISABLE_BOOKING', true);
        this.beds = 0;
        this.$store.dispatch('booking/SET_BEDS', 0);
        this.$store.dispatch('booking/SET_AVAILABLE_BEDS', 0);
      } else {
        this.$store.dispatch('booking/SET_NO_DATES_AVAILABLE', false);
        //this.$store.dispatch('avalanche/DISABLE_BOOKING', false);
      }
    },
    updateDisabledDatesInMenu() {
      this.maximumDateTo = getDateString(undefined, this.daysInAdvanceBooking);
      for (let e in this.disabledDates) {
        const day = getDate(this.disabledDates[e].toString());
        // if a disabled date is after the dateFrom that is chosen
        if (day > getDate(this.dateFrom)) {
          // if the disabled date is smaller than what we already have as maxiumumDateTo
          if (day < getDate(this.maximumDateTo)) {
            this.maximumDateTo = day.toISOString().substr(0, 10);
          }
          // handles edge case where a dateFrom is chosen before a disabled date, when dateTo already was after the disabled date
          if (getDate(this.dateTo) > getDate(this.maximumDateTo)) {
            this.dateTo = getDateString(this.dateFrom, 1);
          }
        }
      }
    },
    checkDatesBeyondBookingWindow(): boolean {
      if (getDate(this.dateFrom) > getDate(this.maximumDate)) {
        this.dateFrom = getDateString();
        return true;
      }
      return false;
    }
  }
});
</script>

<style lang="scss" module>
.row {
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: flex-start;
  padding-top: 12px;
  margin: auto;
  @media only screen and (max-width: 960px) {
    justify-content: space-between;
  }
}

// Every element in the filter menu
.row > div,
v-menu {
  padding-left: 12px;
  padding-right: 12px;
  @media only screen and (max-width: 600px) {
    padding-left: 0;
    padding-right: 0;
  }
}

.bookingTabletRemovePadding {
  @media only screen and (min-width: 600px) and (max-width: 960px) {
    padding-left: 0 !important;
  }
}

.row > div:last-child {
  padding-right: 0;
}

.rowTextFieldArrival {
  max-width: 200px;
  padding-left: 0 !important;
  @media only screen and (max-width: 600px) {
    max-width: 150px;
    margin-right: 0;
  }
}
.rowTextFieldDeparture {
  max-width: 200px;
  @media only screen and (max-width: 600px) {
    max-width: 150px;
  }
}
.bedsText {
  margin-bottom: 0 !important;
}
.availableBedsText {
  margin-bottom: 0 !important;
  font-weight: 100;
  @media only screen and (min-width: 600px) and (max-width: 960px) {
    font-size: 8px;
  }
}

.divBeds {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
}

.bedsInput {
  max-width: 60px;
  margin-left: 8px !important;
  margin-bottom: 0;
}

.column {
  display: flex;
  flex-direction: column;
}

.templateTextField {
  display: flex;
  flex-direction: column;
  margin: 6%;
}
.errorMsg {
  margin-top: 10px;
}
</style>
