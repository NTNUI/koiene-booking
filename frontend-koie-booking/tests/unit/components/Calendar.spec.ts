import Vue from 'vue';
import Vuetify from 'vuetify';
import Vuex, { Store } from 'vuex';
import { storeConfig } from '@/store';
import { cloneDeep } from 'lodash';
import { getDateString } from '@/utils/dates';

Vue.use(Vuetify);

// Components or views
import Calendar from '../../../src/components/Calendar.vue';
import { createWrapper } from '../utils';
import { RootState } from '@/store/types';

describe('Component Calendar.vue', () => {
  let store: Store<RootState>;
  beforeEach(() => {
    store = new Vuex.Store(cloneDeep(storeConfig));
  });

  it('Matches snapshot', async () => {
    const wrapper = createWrapper(Calendar, {
      propsData: {
        allKoier: false,
        row: false,
      },
    });

    expect(wrapper).toMatchSnapshot();
  });

  it('Departure date later than arrival date', async () => {
    const wrapper: any = createWrapper(Calendar, {
      propsData: {
        allKoier: false,
        row: false,
      },
    });

    // Set dateFrom to one day ahead of the initial dateTo-date
    wrapper.vm.dateFrom = getDateString(undefined, 2);
    // Expect dateTo to be the day after the new dateFrom.
    expect(wrapper.vm.dateTo).toBe(getDateString(undefined, 3));
  });

  it('Cannot choose dates around a disabled date', async () => {
    // Setting the required store values instead of mocking.
    const bedsPerDate: any = {};
    const date1 = getDateString(undefined, 2);
    bedsPerDate[date1] = 0;
    store.commit('koie/setKoieData', {
      koie: {
        beds_available_in_booking_window: bedsPerDate,
      },
    });

    const wrapper: any = createWrapper(Calendar, {
      store,
    });

    expect(wrapper.vm.dateFrom).toBe(getDateString());
    expect(wrapper.vm.maximumDateTo).toBe(date1);
  });

  it('Disabled dates is disabled on the dateFrom (arrival) picker', async () => {
    // Setting the required store values instead of mocking.
    const bedsPerDate: any = {};
    const date1 = getDateString(undefined, 2);
    const date2 = getDateString(undefined, 10);
    bedsPerDate[date1] = 0;
    bedsPerDate[date2] = 0;
    store.commit('koie/setKoieData', {
      koie: {
        beds_available_in_booking_window: bedsPerDate,
      },
    });

    const wrapper: any = createWrapper(Calendar, {
      store,
    });

    expect(wrapper.vm.allowedDatesFrom(wrapper.vm.disabledDates[0])).toBe('');
    expect(wrapper.vm.allowedDatesFrom(wrapper.vm.disabledDates[1])).toBe('');
  });

  it('Koie can be booked maximum x days in advance', async () => {
    const wrapper: any = createWrapper(Calendar, {
      data() {
        return {
          disabledDates: [],
        };
      },
      computed: {
        daysInAdvanceBooking() {
          return 14;
        },
      },
    });

    const days = wrapper.vm.daysInAdvanceBooking;
    const expectedMaxDateTo = getDateString(undefined, days);
    const expectedMaxDateFrom = getDateString(undefined, days - 1);

    expect(wrapper.vm.maximumDate).toBe(expectedMaxDateFrom);
    expect(wrapper.vm.maximumDateTo).toBe(expectedMaxDateTo);
  });

  it('Booking disabled when no available dates', async () => {
    // Setting the required store values instead of mocking.
    const bedsPerDate: any = {};
    const today = getDateString();
    const tomorrow = getDateString(undefined, 1);
    bedsPerDate[today] = 0;
    bedsPerDate[tomorrow] = 0;
    store.commit('koie/setKoieData', {
      koie: { beds_available_in_booking_window: bedsPerDate },
    });

    const wrapper: any = createWrapper(Calendar, {
      store,
      computed: {
        daysInAdvanceBooking() {
          return 2;
        },
      },
    });

    expect(wrapper.vm.noDatesAvailable).toBe(true);
    expect(store.state.avalanche.disableBooking).toBe(true);
  });

  it('Reset #beds when switching to dates with less beds available', async () => {
    // Setting the required store values instead of mocking.
    const bedsPerDate: any = {};
    const today = getDateString();
    const tomorrow = getDateString(undefined, 1);
    bedsPerDate[today] = 11;
    bedsPerDate[tomorrow] = 5;
    store.commit('koie/setKoieData', {
      koie: { beds_available_in_booking_window: bedsPerDate },
    });

    const wrapper: any = createWrapper(Calendar, {
      store,
    });

    wrapper.vm.beds = 11;
    wrapper.vm.setBeds(11);
    wrapper.vm.dateFrom = tomorrow;
    expect(wrapper.vm.beds).toBe(1);
  });
});
