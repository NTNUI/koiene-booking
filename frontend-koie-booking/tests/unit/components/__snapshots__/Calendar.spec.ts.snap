// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component Calendar.vue Matches snapshot 1`] = `
<div class="container vbeds column">
  <div class="v-menu">
    <div class="v-input v-input--is-label-active v-input--is-dirty theme--dark v-text-field">
      <div class="v-input__prepend-outer">
        <div class="v-input__icon v-input__icon--prepend"><i aria-hidden="true" class="v-icon notranslate material-icons theme--dark">event</i></div>
      </div>
      <div class="v-input__control">
        <div class="v-input__slot">
          <div class="v-text-field__slot"><label for="input-10" class="v-label v-label--active theme--dark" style="left: 0px; position: absolute;">Arrival</label><input id="input-10" type="text"></div>
        </div>
        <div class="v-text-field__details">
          <div class="v-messages theme--dark"><span name="message-transition" tag="div" class="v-messages__wrapper"></span></div>
        </div>
      </div>
    </div>
    <!---->
  </div>
  <div class="v-menu">
    <div class="v-input v-input--is-label-active v-input--is-dirty theme--dark v-text-field">
      <div class="v-input__prepend-outer">
        <div class="v-input__icon v-input__icon--prepend"><i aria-hidden="true" class="v-icon notranslate material-icons theme--dark">event</i></div>
      </div>
      <div class="v-input__control">
        <div class="v-input__slot">
          <div class="v-text-field__slot"><label for="input-15" class="v-label v-label--active theme--dark" style="left: 0px; position: absolute;">Departure</label><input id="input-15" type="text"></div>
        </div>
        <div class="v-text-field__details">
          <div class="v-messages theme--dark"><span name="message-transition" tag="div" class="v-messages__wrapper"></span></div>
        </div>
      </div>
    </div>
    <!---->
  </div>
  <div class="divBeds">
    <p vbeds="true" class="bedsText">
      Beds
    </p>
    <p class="availableBedsText">( 0 available )</p>
    <div class="v-input v-input--hide-details v-input--is-label-active v-input--is-dirty v-input--dense theme--dark v-text-field v-text-field--single-line v-text-field--enclosed v-text-field--outlined bedsInput">
      <div class="v-input__control">
        <div class="v-input__slot">
          <fieldset aria-hidden="true">
            <legend><span>â€‹</span></legend>
          </fieldset>
          <div class="v-text-field__slot"><input min="1" max="0" vbeds="true" id="input-19" type="number"></div>
        </div>
      </div>
    </div>
  </div>
  <!---->
</div>
`;
